src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -507,26 +522,19 @@ event::ptr primitive_inst::realloc_if_needed() {
         }
     }
 
-    if (_node->is_type<gather>() || _node->is_type<permute>()) {
-        // For the nodes which can be optimized at runtime, input memory is used as output memory
-        // So there is no need to reallocate output memory
-        if (can_be_optimized())
+    // Clear out memory if if was previously reused, but now primitive can't be optimized
+    if (_node->is_type<gather>() || _node->is_type<permute>() || _node->is_type<reshape>() || _node->is_type<reorder>()) {
+        if (can_be_optimized()) {
+            max_output_layout_size = _deps[0].first->max_output_layout_size;
             return ev;
-        // Clear out memory if if was previously reused, but now primitive can't be optimized
-        if (!can_be_optimized() && _outputs[0] && dep_memory_ptr(0)
-            && _network.get_engine().is_the_same_buffer(dep_memory(0), output_memory(0))) {
+        } else if (_outputs[0] && dep_memory_ptr(0) &&
+                   _network.get_engine().is_the_same_buffer(dep_memory(0), output_memory(0))) {
+            // Clear out memory if if was previously reused, but now primitive can't be optimized
             _outputs[0] = nullptr;
             max_output_layout_size = 0;
         }
     }
 
-    // update layout to ensure that it repsects paddings for correct allocation size
-    if (_node->is_type<kv_cache>() && !_impl_params->can_be_optimized()) {
```

Comment:
Why this part is removed? In PR https://github.com/openvinotoolkit/openvino/pull/22128 I fixed related issue, and I think that shouldn't be removed as shape predictor should use padded size

Reply:
I removed because for kv cache, when can_be_opt = false, padding size is 0. However for general case, padding should be tracked. updated.